theano/gpuarray/c_code/dnn_fwd.c

```diff
@@ -358,53 +395,72 @@ APPLY_SPECIFIC(conv_fwd)(PyGpuArrayObject *input, PyGpuArrayObject *kerns,
             use_cached ? "(cache)": "",
             worksize,
             hashkey.c_str()
-      );
+    );
+    if (!(reuse_algo || use_cached)) {
+        // We have selected an algorithm at runtime.
+        // `t` still contains timing about selection step.
+        fprintf(stderr, "\t(selected %s fwd algo in %g milliseconds)\n", selection_name, theano_clock_to_milliseconds(t));
+        if (n_selections > 1) {
+            fprintf(stderr, "\t(selected %lu fwd algos in %g milliseconds (average: %g milliseconds per selection))\n",
+                    n_selections,
+                    theano_clock_to_milliseconds(total_selection_time),
+                    theano_clock_average_to_milliseconds(total_selection_time, n_selections));
+        }
+    }
   }
 #endif
 
   if (params->choose_once) {
     reuse_algo = 1;
   }
 
-  {
-    gpudata *workspace = 0;
-    /*
-     * This is less than ideal since we need to free it after (which
-     * introduces a synchronization point. But we don't have a module
-     * to place a nice get_work_mem() function in.
-     */
-    if (worksize != 0) {
-      workspace = gpudata_alloc(c->ctx, worksize, NULL, 0, NULL);
-      if (workspace == NULL) {
-        PyErr_SetString(PyExc_RuntimeError, "Could not allocate working memory");
-        cuda_exit(c->ctx);
-        return 1;
-      }
+  gpudata *workspace = 0;
+  /*
+   * This is less than ideal since we need to free it after (which
+   * introduces a synchronization point. But we don't have a module
+   * to place a nice get_work_mem() function in.
+   */
+  if (worksize != 0) {
+    workspace = gpudata_alloc(c->ctx, worksize, NULL, 0, NULL);
+    if (workspace == NULL) {
+      PyErr_SetString(PyExc_RuntimeError, "Could not allocate working memory");
+      cuda_exit(c->ctx);
+      return 1;
     }
+  }
 
-    cuda_wait(input->ga.data, GPUARRAY_CUDA_WAIT_READ);
-    cuda_wait(kerns->ga.data, GPUARRAY_CUDA_WAIT_READ);
-    cuda_wait((*output)->ga.data, GPUARRAY_CUDA_WAIT_WRITE);
-
-    for ( int g = 0; g < groups; g++) {
-      err = cudnnConvolutionForward(
-        params->handle,
-        alpha_p,
-        APPLY_SPECIFIC(input), ((char *)PyGpuArray_DEV_DATA(input)) + input_offset * g,
-        APPLY_SPECIFIC(kerns), ((char *)PyGpuArray_DEV_DATA(kerns)) + kern_offset * g,
-        desc, algo,
-        worksize == 0 ? NULL : *(void **)workspace, worksize,
-        beta_p,
-        APPLY_SPECIFIC(output), ((char *)PyGpuArray_DEV_DATA(*output)) + output_offset * g);
-    }
+  cuda_wait(input->ga.data, GPUARRAY_CUDA_WAIT_READ);
+  cuda_wait(kerns->ga.data, GPUARRAY_CUDA_WAIT_READ);
+  cuda_wait((*output)->ga.data, GPUARRAY_CUDA_WAIT_WRITE);
+
+  #ifdef DEBUG
+  t = theano_clock();
```

Comment:
If the user do not enable CUDA_LAUNCH_BLOCKING=1, it will have wrong results. I see two possible fix:
1) Do as with profile=True and test if CUDA_LAUNCH_BLOCKING=1 when cmodule.debug is True
2) make a sync inside the ifdef.

I prefer the second fix as it won't force the sync of all the computation. This would request a sync before taking the first call to theano_clock and another one before the second call for each pair of call.